graphql_compiler/macros/macro_edge/helpers.py

```diff
@@ -53,6 +57,214 @@ def get_directives_for_ast(ast):
+                # That means this call will also need to make changes and return a new object.
+                made_changes = True
+
+            new_selections.append(new_selection_ast)
+        new_selection_set = SelectionSet(new_selections)
+
+    # Find which name to use, and remove all other tags
+    new_directives = ast.directives
+    tag_name_list = [
+        directive.arguments[0].value.value
+        for directive in ast.directives
+        if directive.name.value == TagDirective.name
+    ]
+    tag_names = set(tag_name_list)
+    if len(tag_name_list) != len(tag_names):
+        raise AssertionError(u'The ast should not contain multiple tags with '
+                             u'the same name. {}'.format(tag_name_list))
+    made_changes = made_changes or len(tag_names) > 1
+    if tag_names:
+        name_to_use = None
+        user_specified_names = tag_names & non_macro_names
+        if len(user_specified_names) == 0:
+            name_to_use = min(tag_names, key=len)
+        elif len(user_specified_names) == 1:
+            name_to_use = next(iter(user_specified_names))
+        else:
+            raise AssertionError(u'Multiple tags on the same field are not allowed: {}'
+                                 .format(user_specified_names))
+        name_change_map.update({
+            name: name_to_use
+            for name in tag_names
+        })
+        new_directives = [
+            directive
+            for directive in ast.directives
+            if (directive.name.value != TagDirective.name or
+                directive.arguments[0].value.value == name_to_use)
+        ]
+
+    if not made_changes:
+        # We didn't change anything, return the original input object.
+        return ast, name_change_map
+
+    new_ast = copy(ast)
+    new_ast.selection_set = new_selection_set
+    new_ast.directives = new_directives
+    return new_ast, name_change_map
+
+
+def replace_tag_names(name_change_map, ast):
+    """Replace tag names that are already in use."""
+    if not isinstance(ast, (Field, InlineFragment, OperationDefinition)):
+        return ast
+
+    made_changes = False
+
+    new_selection_set = None
+    if ast.selection_set is not None:
+        new_selections = []
+        for selection_ast in ast.selection_set.selections:
+            new_selection_ast = replace_tag_names(name_change_map, selection_ast)
+
+            if selection_ast is not new_selection_ast:
+                # Since we did not get the exact same object as the input, changes were made.
+                # That means this call will also need to make changes and return a new object.
+                made_changes = True
+
+            new_selections.append(new_selection_ast)
+        new_selection_set = SelectionSet(new_selections)
+
+    # Rename tag names in @tag and @filter directives, and record if we made changes
+    new_directives = []
+    for directive in ast.directives:
```

Comment:
This `for` loops has a lot going on with little spacing and/or comments. Can we make this more manageable to read? I can imagine that, should someone need to modify something here a year from now, it'll be pretty difficult.